subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/TestEnvironmentIntegrationTest/canRunTestsWithCustomSystemClassLoader/src/test/java/org/gradle/MySystemClassLoader.java

```diff
@@ -2,12 +2,13 @@
 
 import java.net.URL;
 import java.net.URLClassLoader;
-import static junit.framework.Assert.*;
 
 public class MySystemClassLoader extends URLClassLoader {
     public MySystemClassLoader(ClassLoader parent) {
         super(new URL[0], parent);
         // Should be constructed with the default system ClassLoader as root
-        assertEquals(getClass().getClassLoader(), parent);
+        if (!getClass().getClassLoader().equals(parent)) {
```

Comment:
Because `org.junit.jupiter.api.Assertions`'s implementation will cause classloader recursive invocation:

```
Caused by: java.lang.IllegalStateException: recursive invocation
        at java.lang.ClassLoader.initSystemClassLoader(ClassLoader.java:1450)
        at java.lang.ClassLoader.getSystemClassLoader(ClassLoader.java:1436)
        at sun.invoke.util.BytecodeDescriptor.parseMethod(BytecodeDescriptor.java:47)
        at sun.invoke.util.BytecodeDescriptor.parseMethod(BytecodeDescriptor.java:41)
        at java.lang.invoke.MethodType.fromMethodDescriptorString(MethodType.java:1067)
        at java.lang.invoke.BoundMethodHandle$Factory.makeCbmhCtor(BoundMethodHandle.java:818)
        at java.lang.invoke.BoundMethodHandle$Factory.makeCtors(BoundMethodHandle.java:763)
        at java.lang.invoke.BoundMethodHandle$SpeciesData.initForBootstrap(BoundMethodHandle.java:361)
        at java.lang.invoke.BoundMethodHandle$SpeciesData.<clinit>(BoundMethodHandle.java:426)
        at java.lang.invoke.BoundMethodHandle.<clinit>(BoundMethodHandle.java:830)
        at java.lang.invoke.LambdaForm.createIdentityForms(LambdaForm.java:1778)
        at java.lang.invoke.LambdaForm.<clinit>(LambdaForm.java:1833)
        at java.lang.invoke.DirectMethodHandle.makePreparedLambdaForm(DirectMethodHandle.java:222)
        at java.lang.invoke.DirectMethodHandle.preparedLambdaForm(DirectMethodHandle.java:187)
        at java.lang.invoke.DirectMethodHandle.preparedLambdaForm(DirectMethodHandle.java:176)
        at java.lang.invoke.DirectMethodHandle.make(DirectMethodHandle.java:83)
        at java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(MethodHandles.java:1660)
        at java.lang.invoke.MethodHandles$Lookup.getDirectMethodNoSecurityManager(MethodHandles.java:1617)
        at java.lang.invoke.MethodHandles$Lookup.getDirectMethodForConstant(MethodHandles.java:1802)
        at java.lang.invoke.MethodHandles$Lookup.linkMethodHandleConstant(MethodHandles.java:1751)
        at java.lang.invoke.MethodHandleNatives.linkMethodHandleConstant(MethodHandleNatives.java:477)
        at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:35)
```
So I think the simplest solution is to use a handwritten `assertTrue`.

Reply:
Seems `org.junit.jupiter.api.Assertions.assertTrue` implementation will cause classloader recursive invocation:

```
Caused by: java.lang.IllegalStateException: recursive invocation
        at java.lang.ClassLoader.initSystemClassLoader(ClassLoader.java:1450)
        at java.lang.ClassLoader.getSystemClassLoader(ClassLoader.java:1436)
        at sun.invoke.util.BytecodeDescriptor.parseMethod(BytecodeDescriptor.java:47)
        at sun.invoke.util.BytecodeDescriptor.parseMethod(BytecodeDescriptor.java:41)
        at java.lang.invoke.MethodType.fromMethodDescriptorString(MethodType.java:1067)
        at java.lang.invoke.BoundMethodHandle$Factory.makeCbmhCtor(BoundMethodHandle.java:818)
        at java.lang.invoke.BoundMethodHandle$Factory.makeCtors(BoundMethodHandle.java:763)
        at java.lang.invoke.BoundMethodHandle$SpeciesData.initForBootstrap(BoundMethodHandle.java:361)
        at java.lang.invoke.BoundMethodHandle$SpeciesData.<clinit>(BoundMethodHandle.java:426)
        at java.lang.invoke.BoundMethodHandle.<clinit>(BoundMethodHandle.java:830)
        at java.lang.invoke.LambdaForm.createIdentityForms(LambdaForm.java:1778)
        at java.lang.invoke.LambdaForm.<clinit>(LambdaForm.java:1833)
        at java.lang.invoke.DirectMethodHandle.makePreparedLambdaForm(DirectMethodHandle.java:222)
        at java.lang.invoke.DirectMethodHandle.preparedLambdaForm(DirectMethodHandle.java:187)
        at java.lang.invoke.DirectMethodHandle.preparedLambdaForm(DirectMethodHandle.java:176)
        at java.lang.invoke.DirectMethodHandle.make(DirectMethodHandle.java:83)
        at java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(MethodHandles.java:1660)
        at java.lang.invoke.MethodHandles$Lookup.getDirectMethodNoSecurityManager(MethodHandles.java:1617)
        at java.lang.invoke.MethodHandles$Lookup.getDirectMethodForConstant(MethodHandles.java:1802)
        at java.lang.invoke.MethodHandles$Lookup.linkMethodHandleConstant(MethodHandles.java:1751)
        at java.lang.invoke.MethodHandleNatives.linkMethodHandleConstant(MethodHandleNatives.java:477)
        at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:35)
        at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:110)
```

I didn't investigate it deeply since a handwritten `assertTrue` is a simple solution.